### ISSUE TITLE
configuring ToyVpn to connect to

### ISSUE THREAD

**Issue by dovholuk** [marcusrunge/ToyVpn#1]
opening an issue so that we can have a dialog if needed. I tried out your app, things are deploying fine for me. That's a great start!

Did you build / deploy ToyVpn from here https://android.googlesource.com/platform/development/+/master/samples/ToyVpn or from somewhere else?

**Comment by marcusrunge** [comment]
Yes exactly from there, running on a Debian Hyper-V VM. Everyone is more than welcome to improve my example. It would be nice to have such a "Toy"-thing running on Windows than on Linux.

**Comment by marcusrunge** [comment]
Now I have added something which should give us the Bytes from the Buffer to manipulate

**Comment by dovholuk** [comment]
Is there any way you could post the method you use to start ToyVpnServer and the steps you use to test it all out? doesn't have to be super-thorough, only enough to get one started... for example I followed along from the source and started the server on 'tun0' port 8000. I expect you then went and declared a new 'vpn' in settings, chose the ToyVpn provider, gave it a connection name like 'toytest' and then the server name of the ip address where the ToyVpn server is running?  It'd be great to just post a small bootstrapper script of steps you perform to get things "working". Make sense? I've pulled and I'm giving it another go...

**Comment by marcusrunge** [comment]
There is not even one byte reaching the ToyVpn Server, everything beyond authentication fails… Encap/Decap are beeing triggered, but nothing reaches the destination. I have no idea why. Documentation from Microsoft is still lousy and poor to unusable.

**Comment by dovholuk** [comment]
Do you know if dencapsulate is getting called?  After doing a bunch of work on this we've discovered that the buffer that comes back has more than one encapsulated IP packet with it... in other words the API's are assymetric. Encapsulate works "packet by packet" but decapulate works by giving you access to a buffer with 1 or more full IP packets in it..  

I'm not sure this code works 100% any more (still working through things on my own) but here's something that might get you started if you're still interested...

`public void Decapsulate(VpnChannel channel, 
                        VpnPacketBuffer encapBuffer, 
                        VpnPacketBufferList decapsulatedPackets, 
                        VpnPacketBufferList controlPacketsToSend)
{
    try
    {
        var buf = channel.GetVpnReceivePacketBuffer();

        if (encapBuffer.Buffer.Length > buf.Buffer.Capacity)
        {
            LogLine("DROPPING PACKET!!! " + encapBuffer.Buffer.Length 
                  + " bytes. This is bigger than the buffer capacity of: " 
                  + encapBuffer.Buffer.Length, channel);
        }
        else
        {
            byte[] bytes = encapBuffer.Buffer.ToArray();

            int srcOffset = 0;
            int remainingBytes = (int)encapBuffer.Buffer.Length;
            System.IO.Stream stream = encapBuffer.Buffer.AsStream();

            while (remainingBytes > 0)
            {
                //encapBuffer is mashing > 1 packet into the buffer. Read the bytes as an IP and
                //read how long the packet is. Make a packet from those bytes and put them into decapsulate
                IpHeader ipHeader = new IpHeader(stream);
                
                encapBuffer.Buffer.CopyTo((uint)srcOffset, buf.Buffer, 0, ipHeader.TotalLength);
                buf.Buffer.Length = ipHeader.TotalLength;
                decapsulatedPackets.Append(buf);

                srcOffset += ipHeader.TotalLength;
                remainingBytes -= ipHeader.TotalLength;
                
                buf = channel.GetVpnReceivePacketBuffer(); //get a new buffer...
            }

            return;
        }
    }
}`

you also need to parse the bytes as an IP packet. I took inspiration from here https://code.msdn.microsoft.com/How-to-read-IP-header-and-a754677a/sourcecode?fileId=149579&pathId=1849545365 (and then rewrote it for my own uses)

Hope that helps?

**Comment by marcusrunge** [comment]
Thanks, when I have time I will try again. My objective is to connect to a remote TUN/TAP interface. I think this UWP IVpnInterface it not designed for this. But I will try again later… It's going to be a lot of low Level work.

**Comment by dovholuk** [comment]
It sounds like we are on similar paths/trajectories. I have a similar goal and the vpn api as-is, most definitely is not designed for this use case. In order to get what you want to work - you'll need to work around the API and not only run the VPN client but also run a TCP listener. You need to allow the VPN client to 'own' the sending and receiving of payloads. effectively you need to be your own "local vpn server".  I have not found any way to use the VPN api without actually creating a StreamSocket/DatagramSocket

That or you need to leave the UWP/VPN Plugin app and write an NDIS/WFP driver.  If you do get back to it though - I'd love to hear any updates as you go

**Comment by dovholuknf** [comment]
Ever get anything working with the VPN plugin? I picked it up recently - nothing useful has changed and put it back down again... moving on to trying Wintun now.
